# W02A. 이 데이터를 믿을 수 있는가? — 데이터 리터러시와 변수 정의

> 이 데이터는 누가, 왜, 어떻게 모았으며, 어떤 한계가 있을까?

**학습 목표**
1. 데이터의 출처·수집 목적·표본 특성·편향 가능성을 파악하는 **데이터 리터러시** 태도를 설명할 수 있다.
2. 변수의 **척도 수준**(명목/서열/등간/비율)을 구분하고, 척도가 분석 기법 선택에 미치는 영향을 설명할 수 있다.
3. 결측치의 유형(MCAR/MAR/MNAR)을 구분하고, 유형에 따른 처리 방법의 차이를 설명할 수 있다.

## 오늘의 질문/목표/진단

### 오늘의 질문

지난 1주차에서는 GitHub Copilot을 설치하고, 코딩·디버깅·해석문 다듬기 조교로 활용하는 방법을 배웠다. 이제 본격적으로 데이터를 다루기 시작하는데, 그 전에 한 가지 중요한 질문이 있다 — **"이 데이터를 과연 믿을 수 있는가?"** 숫자를 분석하기 전에, 데이터가 어디서 왔고, 누가, 왜, 어떻게 모았는지를 먼저 확인하는 습관이 필요하다. 오늘은 그 습관의 출발점인 **데이터 리터러시**를 배운다.

### 빠른 진단 (1문항)

> **Q.** "한국 대학생 200명을 대상으로 한 설문 결과, 국민의 70%가 현 정부 정책에 만족하는 것으로 나타났다." 이 문장의 문제점은?
>
> ① 설문 대상이 200명으로 너무 적다
> ② 대학생 표본을 국민 전체로 일반화했다
> ③ 70%는 과반이므로 문제없다

## 핵심 개념 학습

### 왜 필요한가

1주차에서 Copilot으로 CSV 파일을 불러오고 간단한 요약을 만들어 보았다. 그런데 코드가 정상적으로 실행되었다고 해서 분석 결과를 바로 믿어도 되는 것은 아니다. **데이터 자체에 문제가 있으면, 아무리 정교한 분석을 해도 결론은 틀릴 수 있다.** 이것을 흔히 "쓰레기가 들어가면 쓰레기가 나온다(Garbage In, Garbage Out)"라고 표현한다.

따라서 분석의 첫 단계는 코드를 실행하는 것이 아니라, **데이터를 의심하는 것**이다. 오늘은 세 가지를 배운다. 첫째, 데이터를 의심하는 태도(데이터 리터러시). 둘째, 변수의 척도를 구분하는 방법. 셋째, 결측치(빠진 값)의 유형과 처리 전략이다.

### 핵심 개념 1: 데이터 리터러시(data literacy)

**데이터 리터러시(data literacy)**란 데이터의 출처·수집 목적·표본 특성·편향 가능성을 파악하는 태도이다. 쉽게 말해서, 숫자를 분석하기 전에 **"이 데이터를 믿을 수 있는가?"**를 먼저 질문하는 것이다.

예를 들어, "이 설문은 대학생만 대상이므로, 결과를 전체 국민에게 일반화할 수 없다"와 같이, 데이터의 한계를 인식하는 것이 데이터 리터러시의 핵심이다. 이 태도는 이후 모든 과제에서 해석문의 "(주의) 한계" 항목을 작성할 때 직접 활용된다.

데이터를 처음 받으면 다음 체크리스트를 확인한다.

**데이터 신뢰 체크리스트:**
- [ ] 수집 주체와 목적이 명확한가?
- [ ] 표본 추출 방법이 적절한가? (무작위 추출 vs 편의 추출)
- [ ] 응답률과 결측 패턴이 보고되었는가?
- [ ] 변수 정의(**코드북**, codebook)가 제공되었는가?

**코드북(codebook)**이란 각 변수의 이름, 정의, 측정 방법, 범주 코딩 규칙 등을 정리한 문서이다. 쉽게 말해서, "이 숫자가 무슨 뜻인가?"를 알려주는 설명서이다. 예를 들어, 성별 변수에서 1=남성, 2=여성이라는 정보가 코드북에 들어 있다. 코드북 없이 데이터를 분석하면 변수의 의미를 정확히 알 수 없으므로, 해석이 불가능하다.

#### 예측 투표 — 척도 판별

아래 변수를 보고, 각각의 척도 수준이 무엇인지 투표로 확인해 보자.

> **투표 질문**: "리커트 5점 만족도 척도(① 매우 불만족 ② 불만족 ③ 보통 ④ 만족 ⑤ 매우 만족)는 어떤 척도인가?"
>
> ① 명목 ② 서열 ③ 등간 ④ 비율
>
> (짧게 생각 후 선택)

많은 학생이 ③ 등간을 선택한다. 그러나 보수적으로 보면, 정답은 **② 서열**이다. "불만족"과 "보통" 사이의 간격이 "보통"과 "만족" 사이의 간격과 동일하다는 보장이 없기 때문이다. 실제 연구에서는 편의상 등간으로 취급하는 경우도 있지만, **이 선택에는 항상 가정이 따른다**는 점을 인식해야 한다.

### 핵심 개념 2: 변수의 척도 수준(level of measurement)

**변수의 척도 수준(level of measurement)**은 변수가 담고 있는 정보의 종류와 수준을 나타낸다. 척도에 따라 사용할 수 있는 분석 기법이 달라지므로, 분석 전에 반드시 확인해야 한다.

| 척도 | 특성 | 예시 | 가능한 연산 |
|---|---|---|---|
| **명목(nominal)** | 이름/범주만 구분 | 성별, 지역, 종교 | 빈도, 최빈값 |
| **서열(ordinal)** | 순서가 있음 | 교육수준, 만족도 등급 | 빈도, 중앙값, 순위 |
| **등간(interval)** | 간격이 동일, 영점이 임의 | 온도(섭씨), IQ | 평균, 표준편차 |
| **비율(ratio)** | 절대 영점이 존재 | 소득, 연령, 체중 | 평균, 비율 비교 |

척도를 판별할 때는 다음 **의사결정 트리**를 활용한다.

**표 W02.1** 척도 구분 의사결정 가이드

| 질문 | 예 → | 아니오 → |
|---|---|---|
| 범주(카테고리)인가? | 명목 or 서열 | 등간 or 비율 |
| 순서가 있는가? | 서열 | 명목 |
| 간격이 동일한가? | 등간 or 비율 | 서열 |
| 절대 영점이 있는가? | 비율 | 등간 |

**척도가 분석 기법 선택을 결정한다.** 예를 들어, 명목 변수끼리의 관계는 카이자승 검정(4주차)으로, 등간/비율 변수의 집단 간 평균 차이는 t-검정(6주차)으로 분석한다. 척도를 잘못 판단하면 부적절한 분석 기법을 선택하게 되며, 그 결과는 신뢰할 수 없다.

### 코딩(부호화): 설문지에서 데이터로

실제 사회조사에서 데이터는 대부분 설문지에서 출발한다. 설문 응답을 숫자로 변환하는 과정을 **코딩(coding)** 또는 **부호화**라고 한다.

흐름은 다음과 같다.

> 설문지 작성 → **코드북** 정의(변수명, 척도, 코딩 규칙) → 응답을 숫자로 변환 → CSV 파일 생성

예를 들어, 설문지에 "귀하의 최종 학력은?"이라는 문항이 있고, 보기가 "① 고졸이하 ② 대졸 ③ 대학원이상"이라면, 코드북에는 다음과 같이 기록한다.

| 변수명 | 척도 | 코딩 |
|---|---|---|
| 교육수준 | 서열 | 1=고졸이하, 2=대졸, 3=대학원이상 |

이때 숫자 1, 2, 3은 **순서**를 나타낼 뿐이다. 1과 2 사이의 간격이 2와 3 사이의 간격과 같다는 뜻이 아니다. 이 점을 잊고 평균을 구하면 해석이 어색해진다("평균 교육수준이 1.8이다"는 무슨 뜻인가?).

### 핵심 개념 3: 결측치(missing data)

**결측치(missing data)**란 데이터에서 값이 비어 있는 경우를 말한다. 설문에서 응답하지 않은 문항, 입력 오류, 시스템 장애 등 다양한 원인으로 발생한다. 결측치를 어떻게 처리하느냐에 따라 분석 결과가 달라질 수 있으므로, 결측의 **유형**을 먼저 파악해야 한다.

**표 W02.2** 결측치 유형별 처리 전략

| 유형 | 정의 | 처리 방법 | 예시 |
|---|---|---|---|
| **MCAR**(완전 무작위 결측) | 결측이 어떤 변수와도 무관 | 삭제 가능 | 설문지 분실, 입력 실수 |
| **MAR**(무작위 결측) | 결측이 다른 관찰된 변수에 의존 | 다중 대체(MI) 권장 | 고연령자가 SNS 이용시간 문항 미응답이 많음 |
| **MNAR**(비무작위 결측) | 결측이 결측값 자체에 의존 | 모델 기반 접근 필요 | 고소득자가 소득 미응답 |

쉽게 말해서, MCAR은 "아무 이유 없이 빠진 것", MAR은 "다른 변수와 관련 있게 빠진 것", MNAR은 "빠진 값 자체와 관련 있게 빠진 것"이다.

**핵심 원칙**: 결측이 있는 행을 무조건 삭제하는 것은 위험하다. 결측 비율이 높거나(20% 이상), 결측 패턴이 무작위가 아닌 경우에는 삭제 대신 **대체(imputation)**를 고려해야 한다. 결측치를 무시하고 분석하면, 표본이 편향되어 결론이 왜곡될 수 있다.

#### 짝 점검 — 변수 척도 분류

아래 5개 변수의 척도를 각자 1분간 분류한 후, 옆 사람과 비교한다. 다르게 분류한 항목이 있으면 30초간 토론한다.

| 변수 | 나의 분류 | 짝의 분류 | 합의 |
|---|---|---|---|
| 성별(남/여) | ___ | ___ | ___ |
| 연령(만 나이) | ___ | ___ | ___ |
| 교육수준(고졸이하/대졸/대학원이상) | ___ | ___ | ___ |
| 월 소득(만 원) | ___ | ___ | ___ |
| 생활만족도(5점 리커트) | ___ | ___ | ___ |

### 예시: 가상 사회 설문 데이터 구조

아래는 이번 실습에서 사용할 데이터(social_survey_w02.csv)의 변수 구성이다.

**표 W02.3** 실습 데이터 변수 구성

| 변수명 | 설명 | 척도 | 예시 값 |
|---|---|---|---|
| 성별 | 응답자 성별 | 명목 | 남성, 여성 |
| 연령 | 만 나이 | 비율 | 21, 35, 48 |
| 교육수준 | 최종 학력 | 서열 | 고졸이하, 대졸, 대학원이상 |
| 소득 | 월 소득(만 원) | 비율 | 150, 300, 500 |
| 생활만족도 | 5점 리커트 | 서열 | 1, 2, 3, 4, 5 |
| 정치관심도 | 5점 리커트 | 서열 | 1, 2, 3, 4, 5 |

이 데이터의 코드북을 확인하고 각 변수의 척도를 판별하는 것이 오늘 실습의 첫 번째 과제이다.

## 실습 과제 (수업 내 완결형)

### 과제 개요

오늘 실습에서는 제공된 데이터의 변수 정보와 결측치 현황을 확인하고, 각 변수의 척도를 판별한다. 1주차에서 배운 Copilot을 활용하여 결측치 유형을 판단하는 연습도 함께 수행한다. 이 주차에서는 해석문 4단계 구조를 본격 도입하지 않으므로, 간략한 형태로 서술한다.

### 기본(Basic) 트랙

#### ① 실행

제공된 데이터의 변수 목록을 출력하고, 각 변수의 척도를 분류표로 작성한다.

```python
import pandas as pd

df = pd.read_csv("social_survey_w02.csv")

# 1. 변수 정보 확인
print("=== 변수 정보 ===")
print(df.info())

# 2. 결측치 현황
print("\n=== 결측치 현황 ===")
print(df.isnull().sum())

# 3. 변수별 고유값 수 확인 (척도 판별 참고)
print("\n=== 변수별 고유값 ===")
for col in df.columns:
    print(f"{col}: {df[col].nunique()}개 고유값 — {df[col].dtype}")
```

_전체 코드는 practice/W02/code/W02-1-data-literacy.py 참고_

**완료 체크**: 변수 목록과 결측치 현황이 출력되었는가? 각 변수의 고유값 수가 확인되었는가?

#### ② 확인

결측치 현황을 확인하고, Copilot 응답을 확인하고 검증한다.

**2-1. 결측치 비율 확인**: 변수별 결측 비율을 계산한다.

```python
print("\n=== 변수별 결측 비율 ===")
print((df.isnull().mean() * 100).round(1))
```

**2-2. Copilot 프롬프트 (필수 1)**: 결측 비율이 가장 높은 변수에 대해 Copilot에게 다음을 질문한다.

> "이 변수의 결측치가 MCAR/MAR/MNAR 중 어디에 해당할 가능성이 높을까? 각 유형일 경우 어떻게 처리해야 하는지 설명해줘."

Copilot의 답변을 읽고, **1가지를 채택 또는 거절**한다. 채택/거절 이유를 결과 근거로 2문장 이내로 작성한다.

*(예: "Copilot이 '소득 변수의 결측은 MNAR일 가능성이 있다'고 답했다. 채택한다. 고소득자가 소득 공개를 꺼려 미응답할 가능성이 있으며, 이 경우 결측값 자체에 결측 여부가 의존하기 때문이다.")*

**완료 체크**: 결측 비율이 가장 높은 변수를 확인했는가? Copilot 채택/거절 기록이 있는가?

#### ③ 결측치 대체(imputation) 실습

앞에서 결측 현황을 확인했다. 이제 결측이 MCAR(완전 무작위 결측)이라고 가정하고, 실제로 결측을 **보완(대체)**하여 전후 차이를 체감한다. MAR·MNAR은 처리 방법이 복잡하므로 이 수업에서는 다루지 않는다.

**3-1. 대체 전 현황 확인**: 결측이 있는 변수의 기술통계를 저장한다.

```python
# 대체 전 소득 통계 저장
before_mean = df["소득"].mean()
before_std = df["소득"].std()
n_missing = df["소득"].isnull().sum()

print(f"소득 결측: {n_missing}건 ({df['소득'].isnull().mean()*100:.1f}%)")
print(f"대체 전 — 평균: {before_mean:.1f}, 표준편차: {before_std:.1f}")
```

**3-2. 결측치 보완 (중앙값 대체)**: 수치형 변수의 결측을 중앙값으로, 범주형 변수의 결측을 최빈값으로 대체한다.

```python
df_filled = df.copy()

# 수치형: 중앙값 대체
for col in ["소득", "생활만족도"]:
    median_val = df_filled[col].median()
    df_filled[col] = df_filled[col].fillna(median_val)
    print(f"{col}: 중앙값({median_val:.0f})으로 대체")

# 범주형: 최빈값 대체
for col in ["성별", "교육수준"]:
    mode_val = df_filled[col].mode()[0]
    df_filled[col] = df_filled[col].fillna(mode_val)
    print(f"{col}: 최빈값('{mode_val}')으로 대체")

print(f"\n남은 결측: {df_filled.isnull().sum().sum()}건")
```

**3-3. 대체 전후 비교**: 소득 변수의 평균·표준편차가 어떻게 변했는지 비교한다.

```python
after_mean = df_filled["소득"].mean()
after_std = df_filled["소득"].std()

print("\n=== 소득 변수 전후 비교 ===")
print(f"{'':>12} {'평균':>10} {'표준편차':>10}")
print(f"{'대체 전':>12} {before_mean:>10.1f} {before_std:>10.1f}")
print(f"{'대체 후':>12} {after_mean:>10.1f} {after_std:>10.1f}")
```

**3-4. 해석 질문**: 중앙값 대체 후 표준편차가 줄어드는 이유를 1문장으로 설명하라.

*(힌트: 결측값을 모두 "중심값" 하나로 채우면 데이터가 중앙으로 몰리기 때문이다.)*

**완료 체크**: 결측 현황 확인 → 중앙값/최빈값 대체 → 전후 비교가 완료되었는가?

_전체 코드는 practice/W02/code/W02-3-mcar-imputation.py 참고_

#### ④ 해석

A–E 형식으로 데이터의 특성을 서술한다.

**제출물 구성:**

- **(A) 분석 질문** (1–2문장): "이 데이터의 변수 구성, 척도, 결측 현황은 어떠한가?"
- **(B) 사용 변수/수준**: 각 변수의 이름과 척도(명목/서열/등간/비율) 기재
- **(C) 결과 근거**: 변수별 결측 현황 + 결측 비율이 가장 높은 변수 명시 + 고유값 수
- **(D) 해석문** (5–7문장): 아래 간략한 구조를 참고하여 직접 작성한다.
- **(E) Copilot 사용 요약** (2–3줄): 어떤 질문을 했고, 어떤 답변을 받았으며, 무엇을 채택/거절했는가?

### 확장(Plus) 트랙

기본 트랙을 완료한 학생은 추가로 다음을 수행한다.

> "이 데이터의 표본이 모집단을 대표하는지 판단하기 위해 어떤 정보가 추가로 필요한가?" 3가지를 서술하라.
>
> *(예: 표본 추출 방법(무작위 vs 편의), 응답률, 모집단의 인구학적 분포 등)*

### Copilot 프롬프트

| 구분 | 프롬프트 |
|---|---|
| 필수 1 | "이 데이터의 변수 '교육수준'은 명목/서열/등간/비율 중 어디에 해당할까? 이유도 설명해줘." |
| 필수 2 | "결측치가 20% 이상인 변수를 어떻게 처리해야 할까? 장단점을 비교해줘." |
| 선택 | "이 데이터의 표본 특성이 모집단 대표성을 가지는지 판단하는 방법을 알려줘." |

- 필수 프롬프트 2개는 실습 중 반드시 사용하고, 결과를 (E)항에 기록한다.
- 선택 프롬프트는 확장 트랙을 수행할 때 활용한다.

## 정리/오류 패턴 공유 + 비판적 읽기

### 흔한 오류 3개

실습 중 자주 나타나는 오류를 정리한다. 자신의 과제에 해당하는 항목이 없는지 점검해 보자.

**오류 1: 척도 혼동 — 리커트 척도를 등간으로 취급**

리커트 5점 척도(매우 불만족~매우 만족)를 등간 척도로 간주하고 평균을 구하는 경우가 많다. 그러나 "불만족"과 "보통" 사이의 심리적 간격이 "보통"과 "만족" 사이의 간격과 동일하다는 보장이 없다. 보수적으로는 **서열 척도**로 취급하며, 등간으로 취급할 경우에는 그 가정을 명시해야 한다.

```python
# 서열 변수에 평균을 구하는 것이 적절한지 생각해 보자
print(df["생활만족도"].mean())  # 이 값의 해석이 타당한가?
print(df["생활만족도"].median())  # 서열 변수에는 중앙값이 더 적절
```

**오류 2: 결측치를 무조건 삭제**

결측이 있는 행을 무조건 삭제(리스트와이즈 삭제)하면, 결측이 무작위가 아닌 경우(MAR, MNAR) 표본이 편향될 수 있다. 결측 비율이 20% 이상이면 **대체(imputation)**를 고려해야 한다.

```python
# 무조건 삭제 (주의 필요)
df_clean = df.dropna()
print(f"원본: {len(df)}행 → 삭제 후: {len(df_clean)}행")
# 얼마나 많은 데이터가 사라졌는지 반드시 확인
```

**오류 3: 코드북 없이 분석 시작**

변수의 의미를 모른 채 분석을 시작하면, 결과를 해석할 수 없다. 예를 들어, 성별 변수에서 1이 남성인지 여성인지 모르면, "1이 더 많다"는 말은 아무 의미가 없다. **코드북 확인이 분석의 첫 단계**이다.

### 해석문 간략 구조

이 주차에서는 해석문 4단계 구조를 본격적으로 도입하지 않는다(3주차에서 정식 도입). 대신, 오늘 과제의 (D)항은 다음 간략한 구조를 따른다.

> **(분석목적)** 데이터의 기본 특성을 파악하기 위해 변수 정보와 결측 현황을 확인했다.
>
> **(핵심근거)** ___개 변수 중 ___개에서 결측이 발견되었으며, '소득' 변수의 결측 비율이 ___% 로 가장 높다. 각 변수의 척도는 명목 ___개, 서열 ___개, 비율 ___개이다.
>
> **(결론)** 분석 전에 '소득' 변수의 결측 처리 방법을 결정해야 하며, 서열 변수에 대해 평균 대신 중앙값 사용을 고려한다.
>
> **(주의)** 결측이 MNAR일 가능성(고소득자 미응답)을 고려해야 하며, 무조건 삭제는 표본 편향을 초래할 수 있다. 이 데이터의 표본이 모집단을 대표하는지 확인하기 위한 추가 정보가 필요하다.

### 비판적 읽기 (5–10분)

아래는 가상의 뉴스 기사 발췌문이다. 읽고 문제점을 찾아보자.

> **〈○○일보〉 2026.03.10**
>
> *"○○대학교 학생 300명을 대상으로 한 설문 결과, 한국 청년의 82%가 정치에 무관심한 것으로 나타났다."*

> **토론 질문**: 이 기사의 해석에서 문제점을 찾아보자. (2–3분, 짝 토론 후 전체 공유)

**문제점 예시:**

1. **표본 대표성**: ○○대학교 학생 300명은 "한국 청년" 전체를 대표하지 않는다. 특정 대학, 특정 지역, 특정 학과의 학생일 가능성이 있으며, 이를 전체 청년으로 일반화하는 것은 **과잉 일반화**이다.

2. **변수 정의 불명확**: "정치에 무관심"이라는 개념을 어떻게 측정했는지 보고되지 않았다. 설문 문항이 "정치에 관심이 있다/없다"인지, "정치 뉴스를 주 ___회 이상 본다"인지에 따라 결과가 완전히 달라질 수 있다. 코드북 없이는 "무관심"의 정확한 의미를 알 수 없다.

3. **응답률/결측 미보고**: 300명이 설문에 참여했다고 했지만, 모집단(전체 학생 수)에서 얼마나 응답했는지, 미응답자가 어떤 특성을 가졌는지 보고되지 않았다. 정치에 관심이 없는 학생이 오히려 설문에 응답하지 않았을 수 있다(MNAR).

## Exit Ticket

### Exit Ticket 문항

> **Q.** 변수의 척도(명목/서열/등간/비율)를 왜 분석 전에 확인해야 하는가? 1문장으로 써라.
>
> *(힌트: 척도와 분석 기법의 관계를 생각해 보라)*

**모범 답안 예시**: "척도에 따라 사용할 수 있는 분석 기법(빈도 분석, 평균 비교, 상관/회귀 등)이 달라지므로, 척도를 잘못 판단하면 부적절한 분석을 선택하게 되기 때문이다."

### 다음 시간 예고

이번 주 B회차(W02B)에서는 의도적으로 결측치와 이상치가 포함된 원시 데이터를 조원(2–3인)과 함께 직접 정리(cleaning)하는 조별 과제를 수행한다. 오늘 배운 척도 구분과 결측치 유형 판별을 실전에 적용하고, 정리 전후의 기술통계를 비교하여 처리가 분포에 미치는 영향을 확인한다.

In [ ]:
# 크로스플랫폼 한글 폰트 설정 (macOS / Windows)
import platform
import matplotlib.pyplot as plt

if platform.system() == "Darwin":
    plt.rcParams["font.family"] = "AppleGothic"
elif platform.system() == "Windows":
    plt.rcParams["font.family"] = "Malgun Gothic"
plt.rcParams["axes.unicode_minus"] = False

In [2]:
import pandas as pd

df = pd.read_csv("data/social_survey_w02_raw.csv", encoding='utf-8')

# 1. 변수 정보 확인
print("=== 변수 정보 ===")
print(df.info())

# 2. 결측치 현황
print("\n=== 결측치 현황 ===")
print(df.isnull().sum())

# 3. 변수별 고유값 수 확인 (척도 판별 참고)
print("\n=== 변수별 고유값 ===")
for col in df.columns:
    print(f"{col}: {df[col].nunique()}개 고유값 — {df[col].dtype}")

=== 변수 정보 ===
<class 'pandas.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   성별      195 non-null    str    
 1   연령      200 non-null    float64
 2   교육수준    192 non-null    str    
 3   소득      175 non-null    float64
 4   생활만족도   190 non-null    float64
 5   정치관심도   200 non-null    int64  
dtypes: float64(3), int64(1), str(2)
memory usage: 9.5 KB
None

=== 결측치 현황 ===
성별        5
연령        0
교육수준      8
소득       25
생활만족도    10
정치관심도     0
dtype: int64

=== 변수별 고유값 ===
성별: 2개 고유값 — str
연령: 47개 고유값 — float64
교육수준: 3개 고유값 — str
소득: 149개 고유값 — float64
생활만족도: 5개 고유값 — float64
정치관심도: 5개 고유값 — int64


In [3]:
print("\n=== 변수별 결측 비율 ===")
print((df.isnull().mean() * 100).round(1))


=== 변수별 결측 비율 ===
성별        2.5
연령        0.0
교육수준      4.0
소득       12.5
생활만족도     5.0
정치관심도     0.0
dtype: float64


In [4]:
# 대체 전 소득 통계 저장
before_mean = df["소득"].mean()
before_std = df["소득"].std()
n_missing = df["소득"].isnull().sum()

print(f"소득 결측: {n_missing}건 ({df['소득'].isnull().mean()*100:.1f}%)")
print(f"대체 전 — 평균: {before_mean:.1f}, 표준편차: {before_std:.1f}")

소득 결측: 25건 (12.5%)
대체 전 — 평균: 362.2, 표준편차: 453.3


In [5]:
df_filled = df.copy()

# 수치형: 중앙값 대체
for col in ["소득", "생활만족도"]:
    median_val = df_filled[col].median()
    df_filled[col] = df_filled[col].fillna(median_val)
    print(f"{col}: 중앙값({median_val:.0f})으로 대체")

# 범주형: 최빈값 대체
for col in ["성별", "교육수준"]:
    mode_val = df_filled[col].mode()[0]
    df_filled[col] = df_filled[col].fillna(mode_val)
    print(f"{col}: 최빈값('{mode_val}')으로 대체")

print(f"\n남은 결측: {df_filled.isnull().sum().sum()}건")

소득: 중앙값(307)으로 대체
생활만족도: 중앙값(3)으로 대체
성별: 최빈값('남성')으로 대체
교육수준: 최빈값('대졸')으로 대체

남은 결측: 0건


In [6]:
after_mean = df_filled["소득"].mean()
after_std = df_filled["소득"].std()

print("\n=== 소득 변수 전후 비교 ===")
print(f"{'':>12} {'평균':>10} {'표준편차':>10}")
print(f"{'대체 전':>12} {before_mean:>10.1f} {before_std:>10.1f}")
print(f"{'대체 후':>12} {after_mean:>10.1f} {after_std:>10.1f}")


=== 소득 변수 전후 비교 ===
                     평균       표준편차
        대체 전      362.2      453.3
        대체 후      355.3      424.2


# W02B. 데이터 정리 체크포인트 — 과제 2

> 원시 데이터의 결측치와 이상치를 어떻게 발견하고 처리하는가?

**학습 목표**
1. Copilot(바이브코딩)을 활용하여 원시 데이터의 결측치와 이상치를 **탐색·처리**할 수 있다.
2. 결측치 처리 전후의 기술통계를 비교하여 처리가 분포에 미치는 영향을 **서술**할 수 있다.
3. 처리 과정과 결과를 코드와 함께 **과제 보고서로 정리**하여 클래스룸에 제출할 수 있다.

## 과제 안내 + 빠른 진단

### 빠른 진단 (1문항, 조별 토론)

> **Q.** 조원끼리 1분간 토론하여 답하라: "결측치가 있는 행을 무조건 삭제하면 어떤 문제가 생길 수 있는가?"

*(정답은 수업 정리 단계에서 정답 예시를 확인한다.)*

## 데이터·바이브코딩 안내

### 오늘의 데이터

**social_survey_w02_raw.csv** — 200명의 가상 사회 설문 데이터이다. 의도적으로 결측치와 이상치가 포함되어 있다.

**표 W02.4** 실습 데이터 변수 구성 및 의도적 문제

| 변수명 | 척도 | 의도적 문제 |
|---|---|---|
| 성별 | 명목 | 결측 5건 |
| 연령 | 비율 | 이상치 3건(999, -1, 200) |
| 교육수준 | 서열 | 결측 8건 |
| 소득 | 비율 | 결측 25건(12.5%), 이상치 2건 |
| 생활만족도 | 서열 | 결측 10건 |
| 정치관심도 | 서열 | 정상 |

### 바이브코딩이란?

오늘 과제는 **바이브코딩(vibe coding)** 방식으로 수행한다. 바이브코딩이란 Copilot에게 데이터와 분석 목표를 자연어로 설명하고, 코드 생성·실행을 요청하는 방식이다. 절차는 다음과 같다.

1. Copilot에게 분석 목표를 자연어로 설명한다.
2. Copilot이 생성한 코드를 읽고 실행한다.
3. 실행 결과를 확인하고, 필요하면 수정을 요청한다.
4. **최종 코드와 실행 결과를 모두 과제물에 포함**한다.

### 바이브코딩 프롬프트 예시

과제 수행 시 아래 프롬프트를 참고하여 Copilot에게 요청한다.

| 구분 | 프롬프트 |
|---|---|
| 결측 탐색 | "social_survey_w02_raw.csv를 불러와서 변수별 결측치 수와 비율을 확인해줘." |
| 이상치 탐색 | "연령 변수에서 100 이상이거나 0 미만인 이상치를 찾아줘." |
| 처리 | "이상치를 NaN으로 바꾸고, 결측치를 수치형은 중앙값, 범주형은 최빈값으로 대체하는 코드를 짜줘." |
| 전후 비교 | "처리 전후의 기술통계를 비교하는 표를 만들어줘." |

## 과제 수행 (바이브코딩)

### 과제 구성 (3개 과제, 총 40분)

| 과제 | 내용 | 보고서 작성 | 배정 시간 |
|---:|---|---|---:|
| 과제 1 | 결측치·이상치 탐색 + 현황 보고 | 조원 A | 15분 |
| 과제 2 | 결측치·이상치 처리 + 전후 비교 | 조원 B | 15분 |
| 과제 3 | 정리 과정 서술 + A–E 보고서 작성 | 함께 | 10분 |

### 과제 1: 결측치·이상치 탐색 + 현황 보고

**분석 질문**: "이 데이터에는 어떤 결측치와 이상치가 있는가?"

**보고서 작성**: 조원 A

**수행 절차**:

① Copilot에게 다음과 같이 요청한다: "social_survey_w02_raw.csv를 불러와서 변수별 결측치 수와 비율, 수치형 변수의 기술통계(최솟값/최댓값 포함)를 확인하는 코드를 짜줘."

② Copilot이 생성한 코드를 실행하고, 결과를 확인한다.

③ 이상치가 의심되는 변수를 추가로 필터링한다. 예를 들어 "연령 변수에서 100 이상이거나 0 미만인 행을 찾아줘"와 같이 요청한다.

④ 아래 현황 보고표를 채운다.

**과제 1 현황 보고표**:

| 변수명 | 척도 | 결측 수 | 결측 비율 | 이상치 여부 | 이상치 값 |
|---|---|---|---|---|---|
| 성별 | 명목 | ___ | ___% | O / X | |
| 연령 | 비율 | ___ | ___% | O / X | |
| 교육수준 | 서열 | ___ | ___% | O / X | |
| 소득 | 비율 | ___ | ___% | O / X | |
| 생활만족도 | 서열 | ___ | ___% | O / X | |
| 정치관심도 | 서열 | ___ | ___% | O / X | |

**실행한 코드와 출력 결과를 모두 보고서에 포함한다.**

### 과제 2: 결측치·이상치 처리 + 전후 비교

**분석 질문**: "결측치와 이상치를 어떻게 처리하며, 처리 전후에 분포가 어떻게 달라지는가?"

**보고서 작성**: 조원 B

**수행 절차**:

① Copilot에게 다음과 같이 요청한다: "원본 데이터를 복사한 뒤, 이상치를 NaN으로 바꾸고, 결측치를 수치형은 중앙값, 범주형은 최빈값으로 대체하는 코드를 짜줘. 처리 전후의 기술통계를 비교하는 표도 만들어줘."

② Copilot이 생성한 코드를 실행하고, 결과를 확인한다.

③ 아래 전후 비교표를 채운다.

**과제 2 전후 비교표**:

| 항목 | 처리 전 | 처리 후 | 변화 |
|---|---|---|---|
| 연령 평균 | ___ | ___ | ___ |
| 연령 표준편차 | ___ | ___ | ___ |
| 소득 평균 | ___ | ___ | ___ |
| 소득 표준편차 | ___ | ___ | ___ |
| 전체 결측 수 | ___ | ___ | ___ |
| 유효 데이터 수 | ___ | ___ | ___ |

**비교 질문**: "이상치를 제거하고 결측치를 중앙값으로 대체한 후, 평균과 표준편차가 어떻게 변했는가? 왜 그런 변화가 나타났는가?" — 이 질문에 대한 답을 2–3문장으로 기록한다.

**실행한 코드와 출력 결과를 모두 보고서에 포함한다.**

### 과제 3: 정리 과정 서술 + A–E 보고서 작성

과제 1과 과제 2의 결과를 바탕으로, 정리 과정 전체를 서술하고 제출물을 완성한다. 조원 두 명이 **함께 합의**하여 작성한다.

#### ① Copilot 검증

Copilot에게 아래 프롬프트 중 **2개를 사용**하고, 결과를 검증한다.

| 구분 | 프롬프트 |
|---|---|
| 필수 1 | "이 데이터에서 이상치를 탐지하는 방법 3가지를 알려줘." |
| 필수 2 | "결측치를 중앙값으로 대체하면 분포에 어떤 영향이 있는지 설명해줘." |
| 선택 | "이 데이터의 결측 패턴이 MCAR인지 확인하는 방법을 알려줘." |

Copilot의 답변 중 **1가지를 채택 또는 거절**하고, 이유를 결과 근거로 2문장 이내로 기록한다.

#### ② 조별 합의 + A–E 보고서 작성

**조별 제출물 구성 (A–E)**:

- **(A) 분석 질문** (1–2문장): "이 데이터에서 어떤 결측/이상치 문제가 있으며, 어떻게 처리했는가?"
- **(B) 사용 변수/척도**: 각 변수의 이름, 척도(명목/서열/비율), 결측/이상치 현황
- **(C) 결과 근거**: Copilot이 생성한 코드 + 처리 전후 기술통계 비교표 (과제 2 전후 비교표)
- **(D) 해석문** (5–7문장): 아래 구조를 참고하여 직접 작성한다.
- **(E) Copilot 사용 요약** (2–3줄): 어떤 질문을 했고, 어떤 답변을 받았으며, 무엇을 채택/거절했는가?

**해석문(D항) 작성 시 참고 구조** (3주차에서 정식 도입하므로 간략 형태):

> **(분석목적)** 원시 데이터의 결측치·이상치를 확인하고, 분석 가능한 형태로 정리하기 위해 탐색적 점검을 수행했다.
>
> **(핵심근거)** ___개 변수 중 ___개에서 결측이 발견되었으며, '소득' 변수의 결측 비율이 ___% 로 가장 높았다. '연령' 변수에서 범위 밖 값(___건)이 발견되어 NaN으로 처리했다.
>
> **(결론)** 이상치 제거 후 연령의 평균이 ___에서 ___으로 변화했으며, 결측치를 중앙값으로 대체한 후 소득의 표준편차가 ___에서 ___으로 변화했다. 처리 후 남은 결측은 ___건이다.
>
> **(주의)** 중앙값 대체는 분산을 줄이는 효과가 있으므로, 실제 분석 시에는 결측 유형에 따른 적절한 처리 방법을 선택해야 한다. 특히 '소득' 변수의 결측이 ___일 가능성을 고려해야 한다.

## 클래스룸 제출

아래 항목을 **Google Classroom 과제란에 조별 1부** 제출한다. 제출물 상단에 **조원 두 명의 이름**을 기재한다.

| 항목 | 내용 |
|---|---|
| 조원 이름 | (두 명 모두 기재) |
| 과제 1 현황 보고표 + 코드·결과 | 변수별 결측/이상치 현황 |
| 과제 2 전후 비교표 + 코드·결과 | 처리 전후 기술통계 비교 |
| (A) 분석 질문 | 1–2문장 |
| (B) 변수/척도 | 변수명 + 척도 + 결측/이상치 현황 |
| (C) 근거 | Copilot 생성 코드 + 처리 전후 비교표 |
| (D) 해석문 | 5–7문장 (간략 구조) |
| (E) Copilot 사용 요약 | 2–3줄 (요청 내용 + 채택/거절 근거) |

## 조별 발표 + 전체 토론

### 조별 발표

2–3개 조가 자신의 결과를 **1분 내외**로 발표한다. 발표 형식은 다음과 같다.

> "우리 조는 결측치 ___건, 이상치 ___건을 발견했다. 이상치는 ___로 처리하고, 결측치는 ___으로 대체했다. 처리 후 연령 평균이 ___에서 ___으로 변했다."

### 전체 토론

- "이상치를 제거하지 않으면 평균이 얼마나 왜곡되는가?"
- "중앙값 대체와 평균 대체는 어떤 차이가 있는가?"
- "결측 비율이 어느 수준이면 삭제 대신 대체를 고려해야 하는가?"

## 정답 확인 + 오류 점검

*(별도 정답지(02C.md)를 참조한다.)*

## Exit Ticket + 다음 시간 예고

### Exit Ticket (개인)

> **Q.** 오늘 과제에서 가장 주의해야 했던 부분은? ① 이상치 탐지 ② 결측 처리 방법 선택 ③ 전후 비교 확인 ④ 해석문 작성 — 1개를 고르고 이유를 1문장으로 써라.

### 다음 시간 예고

2주 동안 데이터를 "의심하고 정리하는" 과정을 배웠다. 다음 주(3주)에서는 정리된 데이터를 가지고 **"전체 모습"을 파악**한다 — 평균, 중앙값, 표준편차, 히스토그램 등 **기술통계와 시각화**를 배우고, 분석 결과를 글로 표현하는 **해석문 4단계 구조**를 처음 도입한다. 오늘 연습한 간략한 해석 서술이 3주차부터 "(분석목적) → (핵심근거) → (결론) → (주의)" 틀로 체계화된다.

# W02C. 데이터 정리 체크포인트 — 정답지

> **이 문서는 W02B 과제의 정답지이다. 과제 제출 후 결과를 확인하는 데 사용한다.**

## 1. 빠른 진단 정답

**질문**: "결측치가 있는 행을 무조건 삭제하면 어떤 문제가 생길 수 있는가?"

**정답**: 결측이 특정 집단에 편중되어 있으면(**MAR**(Missing At Random) 또는 **MNAR**(Missing Not At Random)), 삭제 시 표본이 편향되어 분석 결과가 왜곡될 수 있다. 또한 결측 비율이 높으면 데이터가 크게 줄어들어 통계적 **검정력**(statistical power)이 약해진다.

**보충 설명**: A회차에서 배운 결측 유형(MCAR/MAR/MNAR)을 복기한다. 결측이 무작위(MCAR)가 아닌 경우, 삭제는 체계적 편향을 초래한다.

## 2. 과제 1 기대 결과: 결측치·이상치 탐색

### 기대 코드 예시

```python
import pandas as pd

df = pd.read_csv("social_survey_w02_raw.csv")

# 결측치 현황
print("=== 결측치 현황 ===")
print(df.isnull().sum())
print(f"\n전체 결측 비율: {df.isnull().mean().mean()*100:.1f}%")

# 이상치 확인 (수치형 변수)
print("\n=== 수치형 변수 기술통계 ===")
print(df.describe())

# 연령 이상치 필터링
print("\n=== 연령 이상치 ===")
print(df[(df["연령"] > 100) | (df["연령"] < 0)])

# 소득 이상치 필터링 (극단값)
print("\n=== 소득 기술통계 ===")
print(df["소득"].describe())
```

### 기대 결과 수치

**현황 보고표 (기대 정답)**:

| 변수명 | 척도 | 결측 수 | 결측 비율 | 이상치 여부 | 이상치 값 |
|---|---|---|---|---|---|
| 성별 | 명목 | 5 | 2.5% | X | |
| 연령 | 비율 | 0 | 0% | O | 999, -1, 200 |
| 교육수준 | 서열 | 8 | 4.0% | X | |
| 소득 | 비율 | 25 | 12.5% | O | 극단값 2건 |
| 생활만족도 | 서열 | 10 | 5.0% | X | |
| 정치관심도 | 서열 | 0 | 0% | X | |

**핵심 확인 포인트**:
- 연령 이상치 3건(999, -1, 200)은 `describe()`의 min/max에서 바로 발견 가능하다.
- 소득 이상치 2건은 극단적으로 높은 값으로, 분포의 꼬리에 위치한다.
- 소득의 결측 비율(12.5%)이 가장 높다. 이 수준이면 단순 삭제보다 대체를 고려해야 한다.

## 3. 과제 2 기대 결과: 결측치·이상치 처리 + 전후 비교

### 기대 코드 예시

```python
import pandas as pd

df = pd.read_csv("social_survey_w02_raw.csv")

# 원본 보관
df_raw = df.copy()

# 처리 전 기술통계 저장
before = df.describe()

# 이상치 처리: 범위 밖 값을 NaN으로 변환
df.loc[df["연령"] > 100, "연령"] = pd.NA
df.loc[df["연령"] < 0, "연령"] = pd.NA
# 소득 이상치도 처리 (극단값을 NaN으로)
# 구체적 기준은 데이터에 따라 다름 — IQR 방법 또는 절대 기준 사용

# 결측치 처리: 수치형은 중앙값, 범주형은 최빈값
df["연령"] = df["연령"].fillna(df["연령"].median())
df["소득"] = df["소득"].fillna(df["소득"].median())
df["생활만족도"] = df["생활만족도"].fillna(df["생활만족도"].median())
df["교육수준"] = df["교육수준"].fillna(df["교육수준"].mode()[0])
df["성별"] = df["성별"].fillna(df["성별"].mode()[0])

# 처리 후 기술통계
after = df.describe()
print("=== 처리 전 ===")
print(before)
print("\n=== 처리 후 ===")
print(after)
print(f"\n남은 결측: {df.isnull().sum().sum()}건")
```

### 기대 결과 해설

**전후 비교표 (기대 방향)**:

| 항목 | 처리 전 | 처리 후 | 변화 방향 |
|---|---|---|---|
| 연령 평균 | 높음 (999 등 포함) | 크게 하락 | 이상치 제거 효과 |
| 연령 표준편차 | 매우 큼 | 크게 감소 | 이상치 제거 효과 |
| 소득 평균 | 이상치 포함 | 약간 하락 | 극단값 제거 효과 |
| 소득 표준편차 | 큼 | 감소 | 이상치 제거 + 중앙값 대체 |
| 전체 결측 수 | 48건 이상 | 0건 | 전부 대체 완료 |
| 유효 데이터 수 | (결측 제외) | 200 (전체) | 행 삭제 없이 대체 |

**핵심 해설**:
- 연령의 이상치(999, -1, 200)를 NaN으로 바꾼 뒤 중앙값으로 대체하면, 평균이 극적으로 하락한다. 이것이 이상치가 평균에 미치는 영향을 보여주는 핵심 포인트이다.
- 중앙값 대체 후 표준편차가 감소하는 이유는, 결측값을 모두 "중심값"으로 채우면 데이터가 중앙으로 몰리기 때문이다.
- 소득 결측 25건(12.5%)을 중앙값으로 대체하면 소득 분포가 실제보다 덜 퍼진 형태가 된다. 이 한계를 학생이 인지하는 것이 중요하다.

## 4. 모범 보고서 (A–E)

### (A) 분석 질문

"이 데이터(social_survey_w02_raw.csv)에서 어떤 결측치와 이상치가 존재하며, 이를 처리한 후 기술통계가 어떻게 변하는가?"

### (B) 사용 변수/척도 + 결측/이상치 현황

| 변수명 | 척도 | 결측 | 이상치 |
|---|---|---|---|
| 성별 | 명목 | 5건(2.5%) | 없음 |
| 연령 | 비율 | 0건 | 3건(999, -1, 200) |
| 교육수준 | 서열 | 8건(4.0%) | 없음 |
| 소득 | 비율 | 25건(12.5%) | 2건(극단값) |
| 생활만족도 | 서열 | 10건(5.0%) | 없음 |
| 정치관심도 | 서열 | 0건 | 없음 |

### (C) 결과 근거

Copilot 생성 코드(과제 1, 과제 2에서 실행한 코드 전문) + 처리 전후 기술통계 비교표(과제 2 전후 비교표)를 포함한다.

### (D) 해석문 (모범 답안)

> **(분석목적)** 원시 데이터의 결측치·이상치를 확인하고, 분석 가능한 형태로 정리하기 위해 탐색적 점검을 수행했다.
>
> **(핵심근거)** 6개 변수 중 4개에서 결측이 발견되었으며, '소득' 변수의 결측 비율이 12.5%로 가장 높았다. '연령' 변수에서 범위 밖 값(999, -1, 200)이 3건 발견되어 NaN으로 처리했다. '소득' 변수에서도 극단적 이상치 2건이 발견되었다.
>
> **(결론)** 이상치 제거 후 연령의 평균이 크게 하락하여, 이상치가 평균을 심하게 왜곡하고 있었음을 확인했다. 결측치를 수치형은 중앙값, 범주형은 최빈값으로 대체한 후 전체 결측은 0건이 되었으며, 소득의 표준편차가 감소했다.
>
> **(주의)** 중앙값 대체는 분산을 줄이는 효과가 있으므로, 실제 분석 시에는 결측 유형에 따른 적절한 처리 방법을 선택해야 한다. 특히 '소득' 변수의 결측이 MNAR(고소득자나 저소득자가 응답을 회피)일 가능성을 고려해야 한다. 중앙값 대체는 간편하지만, 결측이 무작위가 아니면 체계적 편향을 남길 수 있다.

### (E) Copilot 사용 요약 (모범 답안)

> Copilot에게 "결측치를 중앙값으로 대체하면 분포에 어떤 영향이 있는지 설명해줘"라고 요청했다. Copilot은 "중앙값 대체는 분산을 줄인다"고 답했다. 채택한다. 실제로 처리 전후 기술통계를 비교한 결과 소득의 표준편차가 감소하여 Copilot의 설명이 실행 결과와 일치했기 때문이다.

## 5. 흔한 오류 3개 + 확인법

### 오류 1: 이상치를 발견하지 못함

`df.describe()`의 min/max 값을 반드시 확인한다. 연령이 999나 -1인 경우는 명백한 이상치이다. 의심되면 직접 필터링하여 확인한다.

```python
# 연령이 100 이상이거나 0 미만인 행을 찾는다
print(df[(df["연령"] > 100) | (df["연령"] < 0)])
```

**주의**: `describe()`를 실행했다면 min=-1, max=999를 그냥 넘기지 않아야 한다. min과 max를 먼저 확인하는 습관이 중요하다.

### 오류 2: 원본 데이터를 덮어쓰기

처리 과정에서 원본 데이터를 직접 수정하면, 나중에 처리 전후를 비교할 수 없다. 원본은 별도로 보관하고, 처리본은 새 변수나 새 파일로 저장하는 것이 좋다.

```python
# 원본 보관
df_raw = df.copy()
# 처리는 df에 수행
df.loc[df["연령"] > 100, "연령"] = pd.NA
```

**주의**: `before = df.describe()`를 처리 전에 실행하지 않으면 비교 자체가 불가능하다. 처리 전 기술통계를 반드시 먼저 저장한다.

### 오류 3: 처리 전후 비교를 생략

결측치나 이상치를 처리한 뒤 "남은 결측 0건"만 확인하고 끝내는 경우가 있다. 반드시 **평균, 표준편차, 분포가 어떻게 변했는지** 비교해야 한다. 처리가 데이터의 특성을 왜곡하지는 않았는지 점검하는 것이 핵심이다.

**주의**: 결측을 채운 뒤 평균과 표준편차가 어떻게 변했는지 반드시 확인한다. 전후 비교가 없으면 처리의 영향을 알 수 없다.

## 6. 비판적 읽기 자료 + 토론 가이드

### 사례

> *"한 연구에서 소득 변수의 결측치(15%)를 모두 삭제하고 평균 소득을 분석했다. 분석 결과, 이 지역 주민의 평균 소득은 350만 원으로 나타났다."*

### 토론 질문

"이 처리 방법에 어떤 문제가 있을 수 있는가?" — 조별로 2–3분 토론 후 전체 공유한다.

### 문제점

1. **MNAR 가능성**: 고소득자가 소득 공개를 꺼려 미응답했다면, 고소득 데이터가 삭제되어 평균 소득이 실제보다 낮게 추정된다. 반대로 저소득자가 미응답했다면 평균이 높게 추정된다. 결측의 원인을 모르면 삭제의 영향을 판단할 수 없다.

2. **표본 크기 감소**: 15%를 삭제하면 유효 데이터가 줄어들어 추후 통계 검정의 **검정력**(statistical power)이 약해진다. 표본이 작을수록 우연에 의한 변동이 커져서 결과의 신뢰도가 낮아진다.

3. **대안 미검토**: 삭제 외에도 중앙값 대체, **다중 대체**(multiple imputation) 등의 방법이 있다. 이러한 대안을 검토하지 않고 바로 삭제를 선택한 것은 정당화가 필요하다.

## 7. Exit Ticket 모범 답안

**질문**: "오늘 과제에서 가장 주의해야 했던 부분은? ① 이상치 탐지 ② 결측 처리 방법 선택 ③ 전후 비교 확인 ④ 해석문 작성 — 1개 고르고 이유 1문장"

**각 선택지별 타당한 이유 예시**:

| 선택지 | 타당한 이유 예시 |
|---|---|
| ① 이상치 탐지 | "연령에 999, -1 같은 값이 있는지 describe()로 확인하지 않으면 이후 모든 분석이 왜곡되기 때문이다." |
| ② 결측 처리 방법 선택 | "중앙값 대체, 평균 대체, 삭제 중 어떤 것을 고를지가 이후 분포에 직접 영향을 미치므로 가장 중요하다." |
| ③ 전후 비교 확인 | "처리를 했는데 전후 비교를 하지 않으면 처리가 데이터를 어떻게 바꿨는지 모르기 때문이다." |
| ④ 해석문 작성 | "숫자만 나열하면 의미가 없고, 변화의 이유를 문장으로 설명하는 것이 분석의 완성이기 때문이다." |

**참고**: 어떤 선택지를 고르든, 이유가 논리적이면 정답이다. 정답이 하나가 아니므로 자기 생각을 솔직하게 쓰면 된다.

## 8. 해석문 모범 구조

자신의 해석문과 아래 모범 구조를 비교하여 빠진 부분이 없는지 점검한다.

> **(분석목적)** 원시 데이터의 결측치·이상치를 확인하고, 분석 가능한 형태로 정리하기 위해 탐색적 점검을 수행했다.
>
> **(핵심근거)** 6개 변수 중 4개에서 결측이 발견되었으며, '소득' 변수의 결측 비율이 12.5%로 가장 높았다. '연령' 변수에서 범위 밖 값(999, -1, 200)이 3건 발견되어 NaN으로 처리했다.
>
> **(결론)** 이상치 제거 후 연령의 평균이 크게 하락하여, 이상치가 평균을 심하게 왜곡하고 있었음을 확인했다. 결측치를 중앙값으로 대체한 후 소득의 표준편차가 감소했다. 처리 후 남은 결측은 0건이다.
>
> **(주의)** 중앙값 대체는 분산을 줄이는 효과가 있으므로, 실제 분석 시에는 결측 유형에 따른 적절한 처리 방법을 선택해야 한다. 특히 '소득' 변수의 결측이 MNAR일 가능성을 고려해야 한다.

**해석문 점검 기준** (학생 해석문을 평가할 때 사용):
- [ ] 4단계 구조(분석목적 → 핵심근거 → 결론 → 주의)를 갖추었는가?
- [ ] 구체적인 수치(결측 수, 이상치 값, 전후 변화)가 포함되어 있는가?
- [ ] "(주의)" 항에서 처리 방법의 한계를 언급했는가?
- [ ] 5–7문장 분량을 충족하는가?

In [7]:
# 서열 변수에 평균을 구하는 것이 적절한지 생각해 보자
print(df["생활만족도"].mean())  # 이 값의 해석이 타당한가?
print(df["생활만족도"].median())  # 서열 변수에는 중앙값이 더 적절

3.094736842105263
3.0


In [8]:
# 무조건 삭제 (주의 필요)
df_clean = df.dropna()
print(f"원본: {len(df)}행 → 삭제 후: {len(df_clean)}행")
# 얼마나 많은 데이터가 사라졌는지 반드시 확인

원본: 200행 → 삭제 후: 154행
